src/org/jgroups/raft/blocks/LockService.java

```diff
@@ -0,0 +1,808 @@
+			prev = lock.waiters.remove(member) ? WAITING : NONE;
+		}
+		if (prev != NONE) unbind(member, lock);
+		if (address.equals(member)) {
+			notifyListeners(lock.id, prev, NONE, false);
+		} else if (address.equals(holder)) {
+			notifyListeners(lock.id, WAITING, HOLDING, false);
+		}
+		if (log.isTraceEnabled()) {
+			log.trace("[%s] %s unlock %s, prev: %s", address, member, lock.id, prev);
+			if (holder != null)
+				log.trace("[%s] %s lock %s, prev: %s, next: %s", address, holder, lock.id, WAITING, HOLDING);
+		}
+		if (waiters != null) for (UUID waiter : waiters) {
+			unbind(waiter, lock);
+			if (address.equals(waiter)) {
+				notifyListeners(lock.id, WAITING, NONE, false);
+			}
+			if (log.isTraceEnabled()) {
+				log.trace("[%s] %s unlock %s, prev: %s", address, waiter, lock.id, WAITING);
+			}
+		}
+	}
+
+	protected void doReset(List<UUID> members) {
+		Set<UUID> prev = new LinkedHashSet<>(memberLocks.keySet());
+		if (log.isTraceEnabled()) {
+			log.trace("[%s] reset %s to %s", address, prev, members);
+		}
+		for (var id : members) prev.remove(id);
+		for (var id : prev) doUnlock(id, prev);
+	}
+
+	protected void bind(UUID member, LockEntry lock) {
+		memberLocks.computeIfAbsent(member, k -> new LinkedHashSet<>()).add(lock);
+	}
+
+	protected void unbind(UUID member, LockEntry lock) {
+		memberLocks.computeIfPresent(member, (k, v) -> {
+			v.remove(lock); return v.isEmpty() ? null : v;
+		});
+	}
+
+	protected void notifyListeners(long lockId, LockStatus prev, LockStatus curr, boolean force) {
+		if (!force && raft.leader() == null) return;
+		if (prev == null) prev = NONE;
+		if (curr == null) curr = NONE;
+		LockStatus local = curr == NONE ? lockStatus.remove(lockId) : lockStatus.put(lockId, curr);
+		if (prev == curr) {
+			prev = local == null ? NONE : local;
+			if (prev == curr) return;
+		}
+		Mutex mutex = mutexes.get(lockId);
+		if (mutex != null) mutex.onStatusChange(lockId, prev, curr);
+		for (Listener listener : listeners) {
+			try {
+				listener.onStatusChange(lockId, prev, curr);
+			} catch (Throwable e) {
+				log.error("Fail to notify listener, lock: %s, prev: %s, curr: %s", lockId, prev, curr, e);
+			}
+		}
+	}
+
+	protected void handleView(View next) {
+		View prev = this.view; this.view = next;
+		if (log.isTraceEnabled()) {
+			log.trace("[%s] View accepted: %s, prev: %s, leader: %s", address, next, prev, raft.leader());
+		}
+
+		if (prev != null) {
+			int majority = raft.raft().majority();
+			if (prev.size() >= majority && next.size() < majority) { // lost majority
+				// In partition case if majority is still working, it will be forced to unlock by reset command.
+				cleanup();
+			} else if (!next.containsMembers(prev.getMembersRaw()) && raft.isLeader()) { // member left
+				try {
+					reset(next);
+				} catch (Throwable e) {
+					log.error("Fail to send reset command", e);
+				}
+			}
+		}
+	}
+
+	protected void cleanup() {
+		lockStatus.forEach((k, v) -> notifyListeners(k, v, NONE, true));
+	}
```

Comment:
My intention is to make both of them real, that means no matter the member is disconnected or partition into a minority subgroup, it will also be unlocked by reset command immediately or eventually.

Reply:
I think it has been fixed, and I adjust the test as well. This method has been renamed to `resign`, although it notifies listeners not based on what really happened in state machine, but it will immediately or eventually happen in state machine.

Reply:
Hrrrmm, a bit tricky. I believe we would want to be notified in these cases (shutdown, partitioning) but doing so from outside the state machine seems a recipe for much more work and patching corner cases. Would it be easier to handle with dedicated methods? Notifications come from the state machine and the other methods to notify of what just happened.